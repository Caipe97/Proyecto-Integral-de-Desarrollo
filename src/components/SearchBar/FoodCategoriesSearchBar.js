import React, { Component } from 'react';
import './SearchBar.css';
import 'bootstrap/dist/css/bootstrap.min.css';
import DataTable from 'react-data-table-component';
import Combobox from "react-widgets/Combobox";
import "react-widgets/styles.css";
import Add from '../../images/add.png';
import TextField from '@material-ui/core/TextField';

const paginacionOpciones={
  rowsPerPageText: 'Filas por Página',
  rangeSeparatorText: 'de',
  selectAllRowsItem: true,
  selectAllRowsItemText: 'Todos',
  
}
const customStyles = {
	header: {
		style: {
			minHeight: '56px',
      background:'#CCDDE2',
      borderRadius:'17px',
      textShadow: '1px 1px black',
		},
	},
	headRow: {
		style: {
      backgroundColor: '#B6E052'
		},
	},
	headCells: {
		style: {
			'&:not(:last-of-type)': {},
		},
	},
  pagination: {
		style: {
			backgroundColor:'#B6E052'
		},
	},
  row:{
    pointer:{
      backgroundColor:'blue'
    }
  },
};
const conditionalRowStyles = [
    {
      when: row => row.caloriesPerServing > 0,
      style: {
        backgroundColor: '#B6E052',
        color: 'black',
        '&:hover': {
          cursor: 'pointer',
        },
      },
    },
  ];


class FoodCategoriesSearchBar extends Component {
  constructor(props) {
    super(props);
    this.state = {
      busquedaCategoria: '',
      alimentos: [],
      categorias: [],
      categoriasFiltradas: '',
      columnas:[],
      quantity: ''
    };
  }

  onChange = async event => {
    event.persist();
    await this.setState({busquedaNombre: event.target.value});
    this.filtrarElementos();
  }

  validateQuantity = (quantity) => { //cannot be empty and only numbers
    const expression = /^\d+$/
    return expression.test(String(quantity).toLowerCase())
  }

  handleChangeInput = async event => {
    event.persist();
    if(this.validateQuantity(event.target.value)){
      await this.setState({quantity: event.target.value})
    } else {
      await this.setState({quantity: ''})
    }
    this.asignarColumnas();
  }

  asignarColumnas=()=>{
    const columnas = [
      {
        cell:(row) => {
          return(
            <div>
              <img src={Add} alt='agregar' onClick={async () => {
                if(this.state.quantity !== '' && this.state.quantity > 0){
                  let objective = {
                    foodCategoryId: row.foodCategoryId,
                    objectiveCalories: parseInt(this.state.quantity, 10),
                  };
                  this.props.onAddObjectiveToCurrentGoal(objective)
                }
                }} id={row.foodId} style={{width: '20px', height: '20px', cursor: 'pointer', marginLeft: 40}}/>
            </div>
          )
        },
        ignoreRowClick: true,
        allowOverflow: true,
        button: true
      },
      {
        name: 'Categoría',
        selector: row => row.name,
        sortable: true
      }
    ];

    this.setState({columnas: columnas});
  }

  filtrarElementos=()=>{
    let search = this.props.foodCategories.filter(foodCategory => {
      if(foodCategory.foodCategoryId.toString() === this.state.busquedaCategoria.toString()){
        return foodCategory;
      } else {
        return '';
      }
    });
    this.setState({categoriasFiltradas: search});
    if(this.state.categoriasFiltradas.length === 0){
      this.setState({categoriasFiltradas: this.props.foodCategories});
    }
  }

  async componentDidMount(){
    await this.props.onGetAllFoods(this.props.userId);
    await this.props.onGetFoodCategories(this.props.userId);
    let foodCategoriesCopy = [...this.props.foodCategories];
    foodCategoriesCopy.unshift({foodCategoryId: 0, name: 'Ver todas las categorías', userId: null});
    this.setState({
      alimentos: this.props.foods,
      categorias: foodCategoriesCopy,
      categoriasFiltradas: this.props.foodCategories
    });
    this.asignarColumnas();
  }
  
  onChangeComboBox = async event => {
    if(event.foodCategoryId === 0){ //Ver todos los alimentos
      await this.setState({...this.state, busquedaCategoria: ''});
    } else if(event.foodCategoryId === -1){ //Crear un nuevo alimento
      await this.setState({...this.state, busquedaCategoria: ''});
    } else {
      await this.setState({...this.state, busquedaCategoria: event.foodCategoryId});
    }
    this.filtrarElementos();
  }

render(){
  return (
    <div className="table-responsive" style={{backgroundColor:'#B6E052'}}>
      <div className="barraBusquedaNombre" style={{backgroundColor:'#B6E052'}}>

        <div style={{display: 'flex', justifyContent: 'space-between', alignItems: 'center'}}>
          <Combobox
            data={this.state.categorias}
            textField='name'
            onSelect={this.onChangeComboBox}
            groupBy={category => category.userId}
            placeholder= "Categoría"
            renderListGroup={ ({group}) => ( //group es el userId
              <span>
                {group === null ? 'Default' : 'Custom'}
              </span>
            )}
            style={{width: "80%"}}
            />
        </div>
        <div style={{width: '80%', backgroundColor: 'white', display: 'flex', flexDirection: 'row', justifyContent: 'center', alignContent: 'center', alignItems: 'center'}}>
          <p>Ingrese la cantidad de calorías que desea agregar: </p>
          <TextField label="Calorias" name='quantity' type='quantity' style={{width: '100px'}} value={this.state.quantity} onChange={this.handleChangeInput} required/>
        </div>
      </div>
      <DataTable 
        columns={this.state.columnas}
        data={this.state.categoriasFiltradas}
        pagination
        paginationComponentOptions={paginacionOpciones}
        fixedHeader
        fixedHeaderScrollHeight="600px"
        noDataComponent={
          <span>No se encontró ninguna categoría</span>
        }
        conditionalRowStyles={conditionalRowStyles}
        customStyles={customStyles}
        pointer
      />
    </div>
  );
}
}
export default FoodCategoriesSearchBar;