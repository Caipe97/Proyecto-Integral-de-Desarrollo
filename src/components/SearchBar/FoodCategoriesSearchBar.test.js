import React from 'react';
import { shallow } from 'enzyme';
import FoodCategoriesSearchBar from './FoodCategoriesSearchBar';

describe('FoodCategoriesSearchBar tests', () => {
    const historyMock = { push: jest.fn() };
    const mockPropsSignIn = {
        userId: 1, 
        onLogout: jest.fn()};
    let wrapper;
    const mockPropsSignIn2 = {
        userId: 0, 
        row:{
            name: 'Milanesa', recommendedServing: 85, caloriesPerServing: 198, foodCategoryId:1
        },

        onLogout: jest.fn(),
        userId: 1,state : {
            busquedaNombre: '',
            busquedaCategoria: '',
            alimentos: [],
            categorias: [],
            columnas:[],
            cantidadAlimentos: 0,
            cantidadCategorias: 0,
            quantity:10,
          },
            meals:[
                {mealId: 21, name: 'queso', foodsAndQuantity: {quantity: 1, food: {foodId: 1, name: 'Milanesa', recommendedServing: 85, caloriesPerServing: 198, createdAt: '2021-09-15T19:58:04.486Z'}}, dateEaten: '2021-09-09T00:00:00.000Z', userId: 1},
                {mealId: 22, name: 'carne', foodsAndQuantity: {quantity: 1, food: {foodId: 1, name: 'Milanesa', recommendedServing: 85, caloriesPerServing: 198, createdAt: '2021-09-15T19:58:04.486Z'}}, dateEaten: '2021-09-09T00:00:00.000Z', userId: 1},
                {mealId: 23, name: 'jamon', foodsAndQuantity: {quantity: 1, food: {foodId: 1, name: 'Milanesa', recommendedServing: 85, caloriesPerServing: 198, createdAt: '2021-09-15T19:58:04.486Z'}}, dateEaten: '2021-09-09T00:00:00.000Z', userId: 1}
            ],
            meal: {
                FoodList:[
                    {quantity: 1, food: {foodId: 1, name: 'Milanesa', 
                    recommendedServing: 85, caloriesPerServing: 198, 
                    createdAt: '2021-09-15T19:58:04.486Z'}}, 
                ]
            },
            foods: [
                {foodId: 1, name: 'Milanesa', recommendedServing: 85, caloriesPerServing: 198, createdAt: '2021-09-15T19:58:04.486Z'},
                {foodId: 2, name: 'Papa', recommendedServing: 200, caloriesPerServing: 100, createdAt: '2021-09-15T20:02:16.490Z'}
            ],
                foodCategories:[
                        {
                          foodCategoryId: 1,
                          name: "Fruta",
                          createdAt: "2021-09-30T14:09:34.756Z",
                          updatedAt: "2021-09-30T14:09:34.756Z",
                          userId: null
                        },
                        {
                          foodCategoryId: 7,
                          name: "Fideos",
                          createdAt: "2021-09-30T18:58:58.542Z",
                          updatedAt: "2021-09-30T18:58:58.542Z",
                          userId: null
                        }
                      ],
        
        onAddMeal: jest.fn(),
        onGetMealsFromUser:jest.fn(),
        onDeleteMeal: jest.fn(),
        onGetAllFoods:jest.fn(),
        onAddFoodToCurrentMeal: jest.fn(),
        onRemoveFoodFromCurrentMeal:jest.fn(),
        onLogout: jest.fn(),
        onResetCurrentMeal: jest.fn(),
        onUpdateCurrentMeal:jest.fn(),
        onChangeCurrentMealNameAndDateEaten:jest.fn(),    
        onAddCustomFood:jest.fn(), 
        onEditCustomFood:jest.fn(),
        onDeleteCustomFood:jest.fn(),
        onGetFoodCategories: jest.fn(),
        onCreateCategory: jest.fn(),
        onDeleteCategory: jest.fn(),
        onEditCategory: jest.fn(),
        preventDefault:  jest.fn()
    };

    beforeEach(() => {
        wrapper = shallow(<FoodCategoriesSearchBar {...mockPropsSignIn2} />);
         
    })

    it('expect to render FoodCategoriesSearchBar component to location state ', async() => {
        expect(wrapper).toMatchSnapshot();//if
        await wrapper.instance().componentDidMount();
    })

    it('OnChange should because name contains search value',async() => {
        wrapper.find('Combobox').at(0).simulate('onSelect');
    })
})
   