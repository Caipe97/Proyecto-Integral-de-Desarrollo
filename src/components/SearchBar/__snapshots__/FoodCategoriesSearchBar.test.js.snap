// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FoodCategoriesSearchBar tests expect to render FoodCategoriesSearchBar component to location state  1`] = `
<div
  className="table-responsive"
  style={
    Object {
      "backgroundColor": "#B6E052",
    }
  }
>
  <div
    className="barraBusquedaNombre"
    style={
      Object {
        "backgroundColor": "#B6E052",
      }
    }
  >
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "justifyContent": "space-between",
        }
      }
    >
      <Combobox
        data={
          Array [
            Object {
              "foodCategoryId": 0,
              "name": "Ver todas las categorías",
              "userId": null,
            },
            Object {
              "createdAt": "2021-09-30T14:09:34.756Z",
              "foodCategoryId": 1,
              "name": "Fruta",
              "updatedAt": "2021-09-30T14:09:34.756Z",
              "userId": null,
            },
            Object {
              "createdAt": "2021-09-30T18:58:58.542Z",
              "foodCategoryId": 7,
              "name": "Fideos",
              "updatedAt": "2021-09-30T18:58:58.542Z",
              "userId": null,
            },
          ]
        }
        groupBy={[Function]}
        onSelect={[Function]}
        placeholder="Categoría"
        renderListGroup={[Function]}
        style={
          Object {
            "width": "80%",
          }
        }
        textField="name"
      />
    </div>
    <div
      style={
        Object {
          "alignContent": "center",
          "alignItems": "center",
          "backgroundColor": "white",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "center",
          "width": "80%",
        }
      }
    >
      <p>
        Ingrese la cantidad de calorías que desea agregar: 
      </p>
      <WithStyles(ForwardRef(TextField))
        label="Calorias"
        name="quantity"
        onChange={[Function]}
        required={true}
        style={
          Object {
            "width": "100px",
          }
        }
        type="quantity"
        value=""
      />
    </div>
  </div>
  <Memo()
    columns={
      Array [
        Object {
          "allowOverflow": true,
          "button": true,
          "cell": [Function],
          "ignoreRowClick": true,
        },
        Object {
          "name": "Categoría",
          "selector": [Function],
          "sortable": true,
        },
      ]
    }
    conditionalRowStyles={
      Array [
        Object {
          "style": Object {
            "&:hover": Object {
              "cursor": "pointer",
            },
            "backgroundColor": "#B6E052",
            "color": "black",
          },
          "when": [Function],
        },
      ]
    }
    customStyles={
      Object {
        "headCells": Object {
          "style": Object {
            "&:not(:last-of-type)": Object {},
          },
        },
        "headRow": Object {
          "style": Object {
            "backgroundColor": "#B6E052",
          },
        },
        "header": Object {
          "style": Object {
            "background": "#CCDDE2",
            "borderRadius": "17px",
            "minHeight": "56px",
            "textShadow": "1px 1px black",
          },
        },
        "pagination": Object {
          "style": Object {
            "backgroundColor": "#B6E052",
          },
        },
        "row": Object {
          "pointer": Object {
            "backgroundColor": "blue",
          },
        },
      }
    }
    data={
      Array [
        Object {
          "createdAt": "2021-09-30T14:09:34.756Z",
          "foodCategoryId": 1,
          "name": "Fruta",
          "updatedAt": "2021-09-30T14:09:34.756Z",
          "userId": null,
        },
        Object {
          "createdAt": "2021-09-30T18:58:58.542Z",
          "foodCategoryId": 7,
          "name": "Fideos",
          "updatedAt": "2021-09-30T18:58:58.542Z",
          "userId": null,
        },
      ]
    }
    fixedHeader={true}
    fixedHeaderScrollHeight="600px"
    noDataComponent={
      <span>
        No se encontró ninguna categoría
      </span>
    }
    pagination={true}
    paginationComponentOptions={
      Object {
        "rangeSeparatorText": "de",
        "rowsPerPageText": "Filas por Página",
        "selectAllRowsItem": true,
        "selectAllRowsItemText": "Todos",
      }
    }
    pointer={true}
  />
</div>
`;
